bots/lending-liquidator/src/liquidate.ts

```diff
@@ -0,0 +1,163 @@
+        // Run the liquidation strategy to find the most profitable liquidation
+        const potentialLiquidation = liquidationStrategy(
+            interBtcApi,
+            liquidatorBalance,
+            oracleRates,
+            undercollateralizedBorrowers,
+            new Map(marketsArray)
+        ) as [MonetaryAmount<CurrencyExt>, CurrencyExt, AccountId];
+        if (potentialLiquidation) {
+            const [amountToRepay, collateralToLiquidate, borrower] = potentialLiquidation;
+            console.log(`Liquidating ${borrower.toString()} with ${amountToRepay.toHuman()} ${amountToRepay.currency.ticker}, collateral: ${collateralToLiquidate.ticker}`);
+            // Either our liquidation will go through, or someone else's will
+            await Promise.all([
+                waitForEvent(interBtcApi.api, interBtcApi.api.events.loans.ActivatedMarket, 10 * APPROX_BLOCK_TIME_MS),
+                interBtcApi.loans.liquidateBorrowPosition(borrower, amountToRepay.currency, amountToRepay, collateralToLiquidate)
+            ]);
+        }
+        
+        // Add any new foreign assets to `chainAssets`
+        chainAssets = new Set([...Array.from(chainAssets), ...foreignAssets]);
+    }
+    
+    export async function startLiquidator(interBtcApi: InterBtcApi): Promise<void> {
+        if (interBtcApi.account == undefined) {
+            return Promise.reject("No account set for the lending-liquidator");
+        }
+        console.log("Starting lending liquidator...");
+
+        const foreignAssets = await interBtcApi.assetRegistry.getForeignAssets();
+        let nativeCurrency = [interBtcApi.getWrappedCurrency(), interBtcApi.getGovernanceCurrency(), interBtcApi.getRelayChainCurrency()]
+        let chainAssets = new Set([...nativeCurrency, ...foreignAssets]);
+        console.log("Listening to new blocks...");
+        let flagPromiseResolve: () => void;
+        const flagPromise = new Promise<void>((resolve) => flagPromiseResolve = () => { resolve(); })
+        // The block subscription is a Promise that never resolves
+        const subscriptionPromise = new Promise(() => {
+            interBtcApi.api.rpc.chain.subscribeNewHeads(async (header) => {
+                console.log(`Scanning block: #${header.number}`);
+                checkForLiquidations(interBtcApi, chainAssets).catch((reason) => {
+                    if (interBtcApi.api.isConnected) {
+                        console.error(reason);
+                    } else {
+                        // If the provider connection is closed, terminate the bot
+                        // Used in the test to shut the bot down
+                        console.log("Terminating...");
+                        flagPromiseResolve();
+                    }
+
+                });
+            });
+        });
+        await Promise.race([
+            flagPromise,
+            subscriptionPromise
```

Comment:
Out of curiosity: If the block subscription promise doesn't ever resolve, and the flag promise resolves immediately, why are we racing here?

Reply:
Ah, I see it now. Thanks for the explanation!